text/0000-cache-types.md

```diff
@@ -0,0 +1,109 @@
+# Meta
+[meta]: #meta
+- Name: Cache Scope
+- Start Date: 2020-06-26
+- Author(s): [Javier Romero](https://github.com/jromero)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Allow for buildpacks to specify what cache layers can be shared across applications.
+
+# Motivation
+[motivation]: #motivation
+
+In order to provide a better experience, platforms such as `pack` could provide a way to share cache across multiple applications. In particular, this is a frequently requested feature ([pack#340](https://github.com/buildpacks/pack/issues/340), [pack#652](https://github.com/buildpacks/pack/issues/652), [slack](https://buildpacks.slack.com/archives/C94UJCNV6/p1588718688412300), [slack](https://buildpacks.slack.com/archives/CD61YAG69/p1586370336098700)) for similar language family apps.
+
+Use Case:
+```
+Given I build a Java app A
+When I build a Java app B
+Then I expect similar maven dependencies not to have to be downloaded again
+```
+
+In order to solve this use case in a safe manner, the platform needs to know what cache it can share with other applications versus what cache is specific to a single application. This information is primarily known by the buildpacks.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Buildpacks will provide a mechanism to denote a cached layer as something that can be shared by associating specifing a scope of "shared".
+
+- Target persona: buildpack author
+
+There will be two defined scopes: 
+
+- `app` - This cached layer is specific to the application and SHOULD NOT be shared with any other applications.
+- `shared` - This cached layer is specific to the buildpack and MAY be shared across multiple applications.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+#### Platform Implementation
+
+Platforms can decide how to _share_ the "shared" cached layers based on their specific use cases.
+
+In `pack`, it being a local development tool, shared cached layers could be implemented as a single global cache that is shared across all applications. Additional options can be provided to restrict or otherwise configure this sharing.
+
+In a different platform, one can foresee that the cache is shared only within a "project" or an "organization".
+
+#### Lifecycle Implementation
+
+1. The lifecycle will provide additional `-shared-cache-dir` and `-shared-cache-image` options where `-cache-dir` and `-cache-image` could be provided.
+2. Layers marked as `cache = true` and `cache-scope = "shared"` will be stored in either `-shared-cached-dir` or `-shared-cache-image`.
+3. If `-share-cached-dir` or `-share-cached-image` are not provide and a buildpack sets a layer to cache with scope `shared`, then it would be cached to the standard cache location (`-cache-dir`/`-cache-image`).
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
Are there any security concerns? Like, could there be secrets meant to be used on one buildpack (like a Maven `settings.xml`) that accidentally get shared with another less-trusted buildpack? It seems not that big of a problem since a buildpack can already do just about any thing (you shouldn't be using untrusted buildpacks).

Are there any debugging concerns? Could a buildpack with a bug that creates a malformed `.m2` buildpack cause an error in another buildpack? What if two buildpacks intend to share a cache layer, but they are using two different versions of some shared schema?

Reply:
> What if two buildpacks intend to share a cache layer, but they are using two different versions of some shared schema?

By "two buildpacks" do you mean buildpacks with different `id`s? I'd interpreted the `This cached layer is specific to the buildpack` in the "what it is" section as meaning the cache is still buildpack-specific, in which case this scenario can't occur?

That said, I can see a few other problematic shared-cache scenarios:
1. The shared-cache is used by two different major versions of the same buildpack, where the cache contents is not compatible between them (for example a new version of the yarn buildpack defaulting to newer yarn, which has a different yarn cache schema)
2. The shared-cache contents is dependant on app source input (for example yarn version specified in `package.json`) meaning the shared-cache can be incompatible even with the same version of buildpack used across all apps

To help with (1), shared-caches could be made specific not only to the buildpack ID, but the buildpack version too. 

However:
* this wouldn't help with (2), so buildpack authors will still need to think about cache versioning themselves
* non-shared caches (presumably) aren't tied to buildpack version either, so doing this would be inconsistent

...so it making the cache buildpack-version specific doesn't seem worthwhile to me.

Though it does seem like it would be worth warning buildpack authors about general cache risks in the spec? (eg "Buildpacks SHOULD ensure that use of cached layers is forwards and backwards compatible across buildpack and tooling versions".)